evaluator.measure-hapi/src/main/java/org/opencds/cqf/cql/evaluator/measure/r4/R4MeasureProcessor.java

```diff
@@ -267,39 +254,43 @@ public MeasureReport evaluateMeasure(Measure measure, String periodStart, String
                 this.validator.validate(measure, true);
             }
         }
-
-        if (this.measureEvaluationOptions.isThreadedEnabled()
-                && subjectIds.size() > this.measureEvaluationOptions.getThreadedBatchSize()) {
-            return threadedMeasureEvaluate(measure, periodStart, periodEnd, reportType, subjectIds, fhirDal,
-                    contentEndpoint, terminologyEndpoint, dataEndpoint, additionalData);
-        } else {
-            return innerEvaluateMeasure(measure, periodStart, periodEnd, reportType, subjectIds, fhirDal,
-                    contentEndpoint,
-                    terminologyEndpoint, dataEndpoint, additionalData);
-
-        }
-
+       var subjectIterator = subjectIds.iterator();
+        var ids = new ArrayList<String>();
+       int threadBatchSize = this.measureEvaluationOptions.getThreadedBatchSize();
+       List<CompletableFuture<MeasureReport>> futures = new ArrayList<>();
+       while (subjectIterator.hasNext()){
+           ids.add(subjectIterator.next());
+           if (ids.size() % threadBatchSize == 0) {
+               futures.add(runEvaluate(measure, periodStart, periodEnd, reportType, ids, fhirDal,
+                       contentEndpoint, terminologyEndpoint, dataEndpoint, additionalData));
+               ids.clear();
+           }
```

Comment:
Here the the ids list is passed into a future and the ids is cleared. It might happen that ids is cleared before the future is completed. You probably has to create another transient list to pass.